docs/tutorial-part2.rst

```diff
@@ -0,0 +1,550 @@
+
+
+Now we can read this file during application creation:
+
+.. code-block:: python
+
+   def create_app():
+       static_app = StaticApplication(STATIC_PATH)
+       routes = [
+           ("/", home, "home.html"),
+           ("/static", static_app),
+       ]
+
+       config_path = os.path.join(CUR_PATH, "erosion.ini")
+       config = ConfigParser()
+       config.read(config_path)
+
+       host_url = config["erosion"]["host_url"].rstrip("/") + "/"
+       resources = {"host_url": host_url}
+
+       render_factory = AshesRenderFactory(CUR_PATH)
+       return Application(routes, resources=resources, render_factory=render_factory)
+
+
+The application resources are kept as items in a dictionary
+(``resources`` in the example).
+After getting the host URL from the configuration file,
+we put it into this dictionary,
+which then gets registered with the application during application
+instantiation.
+
+Endpoint functions can access application resources
+simply by listing them (their dictionary keys) as parameters:
+
+.. code-block:: python
+
+   def home(host_url):
+       return {"host_url": host_url}
+
+
+Let's apply a similar solution for passing the entries to the template.
+Here's a simple implementation for the storage (file ``storage.py``)
+for saving and retrieving link entries:
+
+.. code-block:: python
+
+   import shelve
+
+
+   class LinkDB:
+       def __init__(self, db_path):
+           self.db_path = db_path
+
+       def get_links(self):
+           with shelve.open(self.db_path) as db:
+               entries = list(db.values())
+           return entries
+
+       def get_link(self, alias):
+           with shelve.open(self.db_path) as db:
+               entry = db.get(alias)
+           return entry
+
+       def add_link(self, *, target_url, alias, expiry_time, max_count):
+           entry = {
+               "target": target_url,
+               "alias": alias,
+               "expiry_time": expiry_time,
+               "max_count": max_count,
+               "count": 0,
+           }
+           with shelve.open(self.db_path) as db:
+               db[alias] = entry
+           return entry
+
+
+It's worth noting that the ``.add_link()`` method
+returns the newly added link.
+
+Now, add an option to the configuration file:
+
+.. code-block:: ini
+
+   [erosion]
+   host_url = http://localhost:5000
+   db_path = erosion.db
+
+
+Next, add the database connection to the application resources:
+
+.. code-block:: python
+   :emphasize-lines: 1, 16, 17
+
+   from storage import LinkDB
+
+
+   def create_app():
+       static_app = StaticApplication(STATIC_PATH)
+       routes = [
+           ("/", home, "home.html"),
+           ("/static", static_app),
+       ]
+
+       config_path = os.path.join(CUR_PATH, "erosion.ini")
+       config = ConfigParser()
+       config.read(config_path)
+
+       host_url = config["erosion"]["host_url"].rstrip('/') + '/'
+       db_path = config["erosion"]["db_path"]
```

Comment:
The db_path, not the config_path, right? I thought that the developer might want to set the config file option as an absolute path.

Reply:
Ah, yes, if it's an absolute path (starting with `/`) we should respect that, but if it's a relative path (like `"links.db"`) then what is it relative to? the current working directory or the config path? I personally like the config path, but can see either for this example.

Reply:
At first I thought the current working directory would be better but you're right. Since it's a setting in the config file, it should be relative to the config file.